hrm-service/migrations/20220324154600-create-CaseSections.js

```diff
@@ -0,0 +1,155 @@
+module.exports = {
+  up: async queryInterface => {
+    await queryInterface.sequelize.query(
+      `CREATE SEQUENCE IF NOT EXISTS public."CaseSections_sectionId_seq"
+        INCREMENT 1
+        START 100000
+        MINVALUE 1
+        MAXVALUE 9223372036854775807
+        CACHE 1;`,
+    );
+    console.log("Sequence 'CaseSections_sectionId_seq' created");
+    await queryInterface.sequelize.query(
+      `ALTER SEQUENCE public."CaseSections_sectionId_seq" OWNER TO hrm`,
+    );
+    console.log("Sequence 'CaseSections_sectionId_seq' ownership altered.");
+    await queryInterface.sequelize.query(`CREATE TABLE IF NOT EXISTS public."CaseSections"
+      (
+          "caseId" integer NOT NULL,
+          "sectionType" text COLLATE pg_catalog."default" NOT NULL,
+          "sectionId" text COLLATE pg_catalog."default" NOT NULL DEFAULT nextval('"CaseSections_sectionId_seq"'::regclass),
+          "createdAt" timestamp with time zone NOT NULL,
+          "createdBy" text COLLATE pg_catalog."default" NOT NULL,
+          "updatedAt" timestamp with time zone,
+          "updatedBy" text COLLATE pg_catalog."default",
+          "sectionTypeSpecificData" jsonb,
+          CONSTRAINT "CaseSections_pkey" PRIMARY KEY ("caseId", "sectionType", "sectionId"),
+          CONSTRAINT "CaseSections_caseId_Case_id_fk" FOREIGN KEY ("caseId")
+              REFERENCES public."Cases" (id) MATCH SIMPLE
+              ON UPDATE CASCADE
+              ON DELETE CASCADE
+      )`);
+    await queryInterface.sequelize.query(
+      `ALTER TABLE IF EXISTS public."CaseSections" OWNER to hrm;`,
+    );
+    await queryInterface.sequelize.query(`
+      CREATE INDEX IF NOT EXISTS "fki_CaseSections_caseId_Case_id_fk" ON public."CaseSections" USING btree
+        ("caseId" ASC NULLS LAST)
+      `);
+    await queryInterface.sequelize.query(`
+      INSERT INTO "CaseSections"
+      SELECT
+      cases.id AS caseId,
+          'note' AS "sectionType",
+          CASE 
+       WHEN (notesWithSameId.duplicates = 1) THEN COALESCE(notes."id", nextval('"CaseSections_sectionId_seq"'::regclass)::text) 
+       ELSE nextval('"CaseSections_sectionId_seq"'::regclass)::text 
```

Comment:
Also, asking out of ignorance: let's say the the `CaseSections_sectionId_seq` is 3 and the id 4 is already used in the table (because `notes."id"` was not null in this statement), will `nextval('"CaseSections_sectionId_seq"'::regclass)::text` "skip" the value 4?

Reply:
The idea is for clients to provide IDs going forward, the sequence is just a fallback for legacy clients that aren't providing them with the rest of the data.

There is an argument for 'starting over', it would make the migration code simpler. The main reason I tried to preserve the IDs was to avoid strange issues that might arise during switchover, where the ID loaded onto the client has been replaced on the backend before the user gets chance to save it.

No, sequences aren't smart enough to skip IDs, the scenario you describe would result in a constraint violation - the sequence isn't intended for long term use, like I said, just for legacy client scenarios - we could remove it in the future

The reason I'm going for client generated IDs for the moment is because of the slightly unusual API we need to support, where all the case sections are specified in a PUT - which could mean some case sections are added, some are updated, and some are deleted (rather than explicitly adding, deleting or modifying individual sections). 

In that scenario the back end has to handle the 'client specified a section with an ID that doesn't exist' either way, because it could happen even if the client is only supposed to provide IDs for updates, the data could still be missing on the backend.

We could reject the whole update in those circumstances, but it seems better to just treat that as an add operation using the ID specified, and if we're doing that, then clients may as well specify the IDs,

Remember, the sectionId doesn't have to be _globally_ unique, it only has to be unique within that case, for sections of the same type, so it shouldn't be tough for clients to generate something unique in that scope (a UUID is complete overkill tbh)